working/macros/api/builders.dart

```diff
@@ -83,9 +83,8 @@ abstract class ClassDeclarationBuilder
 /// The api used by [Macro] to get a [TypeDeclaration] for any given
 /// [TypeAnnotation].
 abstract class TypeIntrospector {
-  /// TODO: Figure out how to deal with `FutureOr<T>`, function types, and
-  /// other non-nominal types.
-  Future<TypeDeclaration> resolve(TypeAnnotation annotation);
+  /// Resolves a [NamedTypeAnnotation] to its declaration.
+  Future<TypeDeclaration> resolve(NamedTypeAnnotation annotation);
```

Comment:
You'll want to keep this accepting TypeAnnotation. We'll end up with two forms for all the structural types:

* `FunctionTypeAnnotation` has fields for the parameter and return types that are of type `TypeAnnotation`. It's an unresolved function type.

* `FunctionType` (better name?) has fields for the parameter and return types of type `TypeDeclaration`. This is a resolved function type. It's still a function type—it's still structural and doesn't refer to a named declaration—but the types within it have been resolved now.

We'll have a similar pair of types for `FutureOrTypeAnnotation` and `FutureOrType`. Maybe `NullableTypeAnnotation` and `NullableType`?

Reply:
The way I was thinking about this is for the `FunctionTypeAnnotation` you would just individually resolve the `TypeAnnotation`s within it - and not resolve the entire thing at once.

Reply:
> What if you want to ask a question like "Is this type a subtype of `int Function(Object?)`?"

The api for this today just takes two `TypeAnnotation` objects - which I know seems a bit weird but the api only exists in phase 2 and beyond, where types are resolvable. The `TypeAnnotation` objects basically need to have come from the compiler originally in order for it to be able to know how to resolve it.

We could instead have resolved versions of _all_ the declaration types, which would then have resolved types, and hang the type comparison apis off of those. But it adds a lot more api surface area.